pkg/kubelet/kuberuntime/kuberuntime_manager.go

```diff
@@ -965,7 +965,7 @@ func (m *kubeGenericRuntimeManager) computePodActions(ctx context.Context, pod *
 		// If container does not exist, or is not running, check whether we
 		// need to restart it.
 		if containerStatus == nil || containerStatus.State != kubecontainer.ContainerStateRunning {
-			if kubecontainer.ShouldContainerBeRestarted(&container, pod, podStatus) {
+			if kubecontainer.ShouldContainerBeRestarted(&container, pod, podStatus, utilfeature.DefaultFeatureGate.Enabled(features.EventedPLEG)) {
```

Comment:
we concluded we can not use feature gates as boolean options, one this graduates this will be a constant, so the argument will be useless

Reply:
For the logic here, the handling of CREATED state of container is tricky. 

`ShouldContainerBeRestarted` is called in `kuberuntime_manager.computePodActions`(add container to ContainersToStart) and `kube_pods.convertToAPIContainerStatuses`(adding Waiting state for Created state container). 

In the failure,  we find mutli-containers running for a single container(defined in spec). This indicates a bug handling this State. The flake also shows that with the evented PLEG, waiting state may have empty reason(which may be a middle state).

I plan to add more logs to show every state change possibilities here to find if this needs a more complex fix(I'm afraid that we need).

Reply:
So far this fix has only reduced the failure rate from a very high rate to a very low state. （IMO）

Reply:
I tried to fix without the FG check in https://github.com/kubernetes/kubernetes/pull/122763.

Reply:
if we know at this point that EventedPLEG will always an optin feature, then we need to start thinking on how users can opt-in, we can have a new flag or we can do a PLEGMode option, where people can say: Normal, Evented, Evented with Fallback ... I do not know, what we need to try to avoid is to fix forward if we don't have very clear the problem